04 - Guides, Workflows, & Courses/Guides/How to add automated tests to your plugin.md

```diff
@@ -1,18 +1,405 @@
+export default class MyPluginLogic {
+
+}
+```
+
+2.  Extract the RibbonIcon callback to a function (see [[How to test plugin code that uses Obsidian APIs#Move logic out to separate files]] for more infos how to extract code)
+
+```ts
+// myplugin.ts
+export default class MyPluginLogic {
+  static ribbonIconCallback() {
+    // Called when the user clicks the icon.
+    new Notice('This is a notice!');
+  }
+}
+```
+
+This still carries a call to the Obsidian API (to `Notice`) that needs to be extracted. To not depend on the Obsidian API in our custom file, we'll instead pass the dependency in from our calling context (which will be `main.ts`). 
+
+3. Pass in Obsidian API depending functionality as function parameter. Please note that `any` is not a good choice for a type - we'll take a look at this problem in [[How to add automated tests to your plugin#Using Obsidian API for typing in extracted business logic|Using Obsidian API for typing in extracted business logic]].
+
+```ts
+// myplugin.ts
+export default class MyPluginLogic {
+  static ribbonIconCallback(notice: any) {
+    // Called when the user clicks the icon.
+    new notice('This is a notice!');
+  }
+}
+```
+
+4. Adjust the callback in `main.ts` to use our extracted function
+
+```ts
+// main.ts
+import MyPluginLogic from "./myplugin";
+
+// [other code ...]
+const ribbonIconEl = this.addRibbonIcon('dice', 'Sample Plugin', (evt: MouseEvent) => {
+	MyPluginLogic.ribbonIconCallback(Notice)
+});
+```
+
+Restart Obsidian to check manually that the ribbon icon is still showing you the notice when clicked.
+
+> [!attention] Correct import path
+> It might happen that Visual Studio Code auto-imports `MyPluginLogic` like this:
+> ```ts
+> import MyPluginLogic from "myplugin";
+> ```
+> Mind the missing `./` in front of the file name. If your code is not working, try adjusting the import to `import MyPluginLogic from "./myplugin";`  
+
+Having the business logic extracted, it's now possible to put `myplugin.ts` under test instead of `main.ts` and avoid mocking obsidian API. 
+
+5. Remove or comment out `__mocks__/obsidian.ts` to make sure the following works without any mocks. 
+6. Replace the content of `example.spec.ts` with:
+
+```ts
+// example.spec.ts
+import MyPluginLogic from "./myplugin";
+
+describe("MyPlugin Tests", () => {
+  it('should create a notice with This is a notice! as text', () => {
+    const mockNotice = jest.fn();
+    MyPluginLogic.ribbonIconCallback(mockNotice)
+
+    expect(mockNotice).toHaveBeenCalledWith("This is a notice!")
+  })
+});
+```
+
+6. Run the test via `npm run test`. It should be green.
+
+Mind that you will not be able to test if you really registered a corresponding ribbonIcon with this approach. Generally, separating business logic from the logic thats necessary to integrate with Obsidian can help with readability and maintainability and might be valuable not only for testing. 
+
+##### Using Obsidian API for typing in extracted business logic
```

Comment:
Inconsistent capitalization. Other times you capitalized `business logic` in headings.